token-upgrade/program/src/processor.rs

```diff
@@ -0,0 +1,186 @@
+    let original_mint_info = next_account_info(account_info_iter)?;
+    let new_escrow_info = next_account_info(account_info_iter)?;
+    let new_account_info = next_account_info(account_info_iter)?;
+    let new_mint_info = next_account_info(account_info_iter)?;
+    let new_transfer_authority_info = next_account_info(account_info_iter)?;
+    let original_token_program = next_account_info(account_info_iter)?;
+    let new_token_program = next_account_info(account_info_iter)?;
+    let original_transfer_authority_info = next_account_info(account_info_iter)?;
+
+    // owner checks
+    check_owner(original_account_info, original_token_program.key)?;
+    check_owner(original_mint_info, original_token_program.key)?;
+    check_owner(new_escrow_info, new_token_program.key)?;
+    check_owner(new_account_info, new_token_program.key)?;
+    check_owner(new_mint_info, new_token_program.key)?;
+
+    // PDA derivation check
+    let (expected_escrow_authority, bump_seed) = get_token_upgrade_authority_address_and_bump_seed(
+        original_mint_info.key,
+        new_mint_info.key,
+        program_id,
+    );
+    let bump_seed = [bump_seed];
+    let authority_seeds = collect_token_upgrade_authority_signer_seeds(
+        original_mint_info.key,
+        new_mint_info.key,
+        &bump_seed,
+    );
+    if expected_escrow_authority != *new_transfer_authority_info.key {
+        msg!(
+            "Expected escrow authority {}, received {}",
+            &expected_escrow_authority,
+            new_transfer_authority_info.key
+        );
+        return Err(TokenUpgradeError::InvalidOwner.into());
+    }
+
+    // pull out these values in a block to drop all data before performing CPIs
+    let (token_amount, original_decimals, new_decimals) = {
+        // check mints are actually mints
+        let original_mint_data = original_mint_info.try_borrow_data()?;
+        let original_mint = StateWithExtensions::<Mint>::unpack(&original_mint_data)?;
+        let new_mint_data = new_mint_info.try_borrow_data()?;
+        let new_mint = StateWithExtensions::<Mint>::unpack(&new_mint_data)?;
+
+        // check accounts are actually accounts
+        let original_account_data = original_account_info.try_borrow_data()?;
+        let original_account = StateWithExtensions::<Account>::unpack(&original_account_data)?;
+        let new_escrow_data = new_escrow_info.try_borrow_data()?;
+        let new_escrow = StateWithExtensions::<Account>::unpack(&new_escrow_data)?;
+        let new_account_data = new_account_info.try_borrow_data()?;
+        let _ = StateWithExtensions::<Account>::unpack(&new_account_data)?;
+
+        let token_amount = original_account.base.amount;
+        if new_escrow.base.amount < token_amount {
+            msg!(
+                "Escrow only has {} tokens, needs at least {}",
+                new_escrow.base.amount,
+                token_amount
+            );
+            return Err(ProgramError::InsufficientFunds);
+        }
```

Comment:
should we enforce `original_mint.base.decimals == new_mint.base.decimals` before here? since we dont provide a setup instruction, theres nothing to prevent the person who sets up the upgrade from making them different, and finding out the hard way when their conversion runs out of tokens partway through. enforcing it prevents any burn of the original from happening, giving the mint controller time to contemplate their words and deeds

i guess there might be a usecase for using this to change decimals, but i think exchanging base amounts 1:1 thereby scaling ui amounts feels... weird, and id expect (as a naive user) the behavior to be the ui amount is the same but the base amount scales. but idk if we want to deal with the math if theres no users for it

Reply:
I was split on this too, and decided to punt by doing less.  This will probably be used more as a reference implementation for token creators to fork, make tiny changes, and deploy as needed.  But for whatever version we deploy, it probably makes more sense to put in a little more protection by ensuring that the mint has the same number of decimals.

Ok, you've convinced me!